tools/dm/manually-handling-sharding-ddl-locks.md

```diff
@@ -0,0 +1,436 @@
+
+The initial table structure is:
+
+```sql
+mysql> SHOW CREATE TABLE shard_db_1.shard_table_1;
++---------------+------------------------------------------+
+| Table         | Create Table                             |
++---------------+------------------------------------------+
+| shard_table_1 | CREATE TABLE `shard_table_1` (
+  `c1` int(11) NOT NULL,
+  PRIMARY KEY (`c1`)
+) ENGINE=InnoDB DEFAULT CHARSET=latin1 |
++---------------+------------------------------------------+
+```
+
+The following DDL operation will be executed on the upstream sharded tables to alter the table structure:
+
+```sql
+ALTER TABLE shard_db_*.shard_table_* ADD COLUMN c2 INT;
+```
+
+The operation processes of MySQL and DM are as follows:
+
+1. The corresponding DDL operations are executed on the two sharded tables of `DM-worker-1` in `MySQL-1` to alter the table structures.
+
+    ```sql
+    ALTER TABLE shard_db_1.shard_table_1 ADD COLUMN c2 INT;
+    ```
+    ```sql
+    ALTER TABLE shard_db_1.shard_table_2 ADD COLUMN c2 INT;
+    ```
+
+2. `DM-worker-1` sends the DDL information related to `MySQL-1` to `DM-master`, and `DM-master` creates the corresponding DDL lock. 
+3. Use `show-ddl-lock` to check the information of the current DDL lock.
+    
+    ```bash
+    » show-ddl-locks test
+    {
+        "result": true,
+        "msg": "",
+        "locks": [
+            {
+                "ID": "test-`shard_db`.`shard_table`",
+                "task": "test",
+                "owner": "127.0.0.1:8262",
+                "DDLs": [
+                    "USE `shard_db`; ALTER TABLE `shard_db`.`shard_table` ADD COLUMN `c2` int(11);"
+                ],
+                "synced": [
+                    "127.0.0.1:8262"
+                ],
+                "unsynced": [
+                    "127.0.0.1:8263"
+                ]
+            }
+        ]
+    }
+    ```
+    
+4. Due to the application demand, the `DM-worker-2` data in `MySQL-2` is no longer needed to be synchronized to the downstream TiDB, and `DM-worker-2` is made offline.
+5. The lock whose ID is ``` test-`shard_db`.`shard_table` ``` on `DM-master` cannot receive the DDL information of `DM-worker-2`.
+
+    - The returned result `unsynced` by `show-ddl-locks` has always included the information of `DM-worker-2` (`127.0.0.1:8263`).
+6. Use `unlock-dll-lock` to ask `DM-master` to actively unlock the DDL lock.
+    - If the owner of the DDL lock has gone offline, you can use the parameter `--owner` to specify another DM-worker as the new owner to execute the DDL.
+    - If any DM-worker reports an error, `result` will be set to `false`, and at this point you should check carefully if the errors of each DM-worker is acceptable and within expectations.
+    - DM-workers that have gone offline will return the error `rpc error: code = Unavailable`, which is within expectations and can be neglected; but if other online DM-workers return errors, then you should deal with them based on the scenario.
```

Comment:
I think it’s example of `If any DM-worker reports an error, `result` will be set to `false`, and at this point you should check carefully if the errors of each DM-worker is acceptable and within expectations.`

Reply:
@GregoryIan Yes, updated.